students/dave_arasim/session07/html_render.py

```diff
@@ -0,0 +1,166 @@
+#!/usr/bin/env python
+
+"""
+A class-based system for rendering html.
+"""
+
+############################################################
+# This is the framework for the base class
+############################################################
+
+class Element(object):
+    tag = 'html'
+    doctype = ''
+    indent_level = 0
+    indent_str = '   '
+
+    def __init__(self, content=None, **kwargs):
+        if content is None:
+            self.contents = []
+        else:
+            self.contents = [content]
+
+        self.attrs = kwargs
+
+    def append(self, new_content):
+        self.contents.append(new_content)
+
+    def _open_tag(self):
+        nv_pairs = ''
+
+        if self.attrs is not None:
+            for key,value in self.attrs.items():
+                nv_pairs += ' ' + str(key) + '='
+                nv_pairs += '"' + str(value) + '"'
+
+        return "<{}{}>".format(self.tag, nv_pairs)
+
+    def _close_tag(self):
+        return "</{}>".format(self.tag)
+
+    def render(self, out_file):
+        if self.doctype != '':
+            out_file.write(self.doctype)
+        out_file.write(Element.indent_str * Element.indent_level)
+        out_file.write(self._open_tag())
+        out_file.write("\n")
+
+        Element.indent_level += 1
```

Comment:
This is an interesting approach and possibly works (unless it's used in multiple threads), but this is essentially like using a global variable to keep track of the indentation level. Probably a better/desired approach is to pass the indentation string (a number of blank spaces) in the `render(...)` call, which will call its sub-element's (content's) `render(...)` with an extended indentation string.